libherokubuildpack/src/buildpack_output/background_printer.rs

```diff
@@ -0,0 +1,195 @@
+//! This module is responsible for the logic involved in the printing to output while
+//! other work is being performed. Such as printing dots while a download is being performed
+use std::io::Write;
+use std::sync::mpsc::{channel, Sender};
+use std::thread::JoinHandle;
+use std::time::Duration;
+
+/// Print dots to the given buffer at the given interval
+///
+/// Returns a struct that allows for manually stopping the timer or will automatically stop
+/// the timer if the guard is dropped. This functionality allows for errors that trigger
+/// an exit of the function to not accidentally have a timer printing in the background
+/// forever.
+#[must_use]
+pub(crate) fn print_interval<W>(
+    mut buffer: W,
+    interval: Duration,
+    start: String,
+    tick: String,
+    end: String,
+) -> PrintGuard<W>
+where
+    W: Write + Send + 'static,
+{
+    let (sender, receiver) = channel::<()>();
+
+    let join_handle = std::thread::spawn(move || {
+        write!(buffer, "{start}").expect("Writer should not be closed");
+        buffer.flush().expect("Writer should not be closed");
+
+        loop {
+            write!(buffer, "{tick}").expect("Writer should not be closed");
+            buffer.flush().expect("Writer should not be closed");
+
+            if receiver.recv_timeout(interval).is_ok() {
+                break;
+            }
+        }
+
+        write!(buffer, "{end}").expect("Writer should not be closed");
+        buffer.flush().expect("Writer should not be closed");
+
+        buffer
```

Comment:
Once the thread is joined it returns the `Write` struct we gave it so we can keep using it.